google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployAction.java

```diff
@@ -120,90 +123,115 @@ public void run() {
+
+    DefaultDeployConfiguration configuration = new DefaultDeployConfiguration();
+    configuration.setDeployables(Collections.singletonList(new File(stagingDirectory, "app.yaml")));
+    configuration.setProject(appEngineHelper.getProjectId());
+    configuration.setPromote(true);
+    if (!StringUtil.isEmpty(version)) {
+      configuration.setVersion(version);
+    }
+
+    deployment.deploy(configuration);
   }
 
-  private class DeployToAppEngineProcessListener extends ProcessAdapter {
-    private StringBuilder deploymentOutput = new StringBuilder();
+  private DefaultProcessRunner createDeployProcessRunner() {
+    final StringBuilder rawDeployOutput = new StringBuilder();
+
+    DefaultProcessRunner processRunner =
+        new DefaultProcessRunner(new ProcessBuilder());
+    processRunner.setAsync(true);
 
-    @Override
-    public void onTextAvailable(ProcessEvent event, Key outputType) {
-      if(outputType.equals(ProcessOutputTypes.STDOUT)) {
-        deploymentOutput.append(event.getText());
+    processRunner.setStdErrLineListener(new ProcessOutputLineListener() {
+      @Override
+      public void outputLine(String output) {
+        consoleLogLn(output);
       }
-    }
+    });
 
-    @Override
-    public void processTerminated(final ProcessEvent event) {
-      try {
-        if (event.getExitCode() == 0) {
-
-          // Parse JSON output to retrieve service/version info.
-          DeployOutput deployOutput = null;
-          try {
-            deployOutput = parseDeployOutput(deploymentOutput.toString());
-          } catch (JsonParseException e) {
-            logger.error("Could not retrieve service/version info of deployed application", e);
-          }
-          // Recommend to update gcloud if we can't get service/version for whatever reasons.
-          if (deployOutput == null
-              || deployOutput.getService() == null || deployOutput.getVersion() == null) {
-            consoleLogLn(GctBundle.message("appengine.deployment.version.extract.failure") + "\n");
-          }
+    processRunner.setStdOutLineListener(new ProcessOutputLineListener() {
+      @Override
+      public void outputLine(String output) {
+        rawDeployOutput.append(output);
+      }
+    });
 
-          callback.succeeded(
-              new AppEngineDeploymentRuntime(
-                  project, appEngineHelper, getLoggingHandler(),
-                  deployOutput != null ? deployOutput.getService() : null,
-                  deployOutput != null ? deployOutput.getVersion() : null));
-        } else if (cancelled) {
-          callback.errorOccurred(GctBundle.message("appengine.deployment.error.cancelled"));
-        } else {
-          logger.warn("Deployment process exited with an error. Exit Code:" + event.getExitCode());
-          callback.errorOccurred(
-              GctBundle.message("appengine.deployment.error.with.code", event.getExitCode()));
+    processRunner.setExitListener(new ProcessExitListener() {
```

Comment:
Yes. Valid point. For deploy, the stop listener would need access to the raw deploy output and I was running into some weird behavior passing in a reference to it (since its assembled by callbacks) so I left it with the intention of getting back to it.  I'll address this since it helps readability to extract it out.